analytics/src/main/java/com/segment/analytics/internal/Utils.java

```diff
@@ -154,7 +161,18 @@ public static float coerceToFloat(Object value, float defaultValue) {
 
     /** Returns true if the application has the given permission. */
     public static boolean hasPermission(Context context, String permission) {
-        return context.checkCallingOrSelfPermission(permission) == PERMISSION_GRANTED;
+        return hasPermission(context, permission, 0);
+    }
+
+    private static boolean hasPermission(Context context, String permission, int repeatCount) {
+        try {
```

Comment:
is it guaranteed to always work on the second try? if so then we should probably just do function calls rather than use recursion here

Reply:
no, it is not guaranteed to work on the second try. This is why it is beneficial to keep it parametrized for future.
This is a known solution, for example you can see:
https://github.com/evernote/android-job/blob/d9cfc62d2cd52051b424efa0dcff3574287a6171/library/src/main/java/com/evernote/android/job/util/JobUtil.java#L87
in the evernote library they do the same.
Basically, if it is failing it fails for a reason of app updating simultaneously with the request for check for permissions. Most likely second invocation will be after package manager updated its state and app update is counted.
If it doesn't fail, then we see result from first try, as we are not going into recursion in case the result is false, only when there is an exception.

Reply:
I see, thanks for the context 👍